### PR TITLE
[7907871] Repo control for branch protection

### PR DESCRIPTION
**Pull request by juhi037**
## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 1]
```diff
@@ -92,6 +96,148 @@ class CommonSVTControls: ADOSVTBase {
...
     }
+    hidden [ControlResult] CheckBuildServiceAccessOnBranch([ControlResult] $controlResult)
+    {
+        $controlResult.VerificationResult = [VerificationResult]::Failed
+        try
+        {
+            $isControlPassing = $true
+            $projectId = ($this.ResourceContext.ResourceId -split "project/")[-1].Split('/')[0]           
+            $url = 'https://dev.azure.com/{0}/_apis/Contribution/HierarchyQuery?api-version=5.0-preview.1' -f $($this.OrganizationContext.OrganizationName);
+            $inputbody = '{"contributionIds":["ms.vss-admin-web.security-view-members-data-provider"],"dataProviderContext":{"properties":{"permissionSetId": "2e9eb7ed-3c0a-47d4-87c1-0ffdd275fd87","permissionSetToken":"","sourcePage":{"url":"","routeId":"ms.vss-admin-web.project-admin-hub-route","routeValues":{"project":"","adminPivot":"repositories","controller":"ContributedPage","action":"Execute"}}}}}' | ConvertFrom-Json
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:107]: keep this "2e9eb7ed-3c0a-47d4-87c1-0ffdd275fd87" at class variable and use that variable here

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 2]
```diff
@@ -92,6 +96,148 @@ class CommonSVTControls: ADOSVTBase {
...
+                            $postbody=@'
+                            {"contributionIds":["ms.vss-admin-web.org-admin-members-data-provider"],"dataProviderContext":{"properties":{"subjectDescriptor":"{0}","sourcePage":{"url":"https://dev.azure.com/{2}/_settings/groups?subjectDescriptor={1}","routeId":"ms.vss-admin-web.collection-admin-hub-route","routeValues":{"adminPivot":"groups","controller":"ContributedPage","action":"Execute"}}}}}
+'@
+                            $postbody=$postbody.Replace("{0}",$broaderGroup.descriptor )
+                            $postbody=$postbody.Replace("{1}",$this.OrganizationContext.OrganizationName)
+                            $rmContext = [ContextHelper]::GetCurrentContext();
+                            $user = "";
+                            $base64AuthInfo = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(("{0}:{1}" -f $user,$rmContext.AccessToken)))   
+                            try {
+                                $response = Invoke-RestMethod -Uri $url -Method Post -ContentType "application/json" -Headers @{Authorization=("Basic {0}" -f $base64AuthInfo)} -Body $postbody
```
Comment (arvindsingh-cloud) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:211]: can we use InvokePostWebRequest? (common method)

File: src/AzSK.ADO/Framework/Configurations/SVT/ADO/ADO.CommonSVTControls.json [hunk 3]
```diff
@@ -38,6 +38,23 @@
...
       "Enabled": true
     },
+    {
+      "ControlID": "ADO_Repository_AuthZ_Dont_Grant_BuildSvc_Permission_On_Branch",
+      "Description": "Do not grant build service groups excessive permissions on repository branches.",
+      "Id": "Repository120",
+      "ControlSeverity": "Medium",
+      "Automated": "Yes",
+      "MethodName": "CheckBuildServiceAccessOnBranch",
+      "Rationale": "If 'Project Collection Build Service' or 'Project Build Service' groups have excessive permissions on important branches of a repository, then a malicious user can access the repository and tamper its contents, bypassing any defined policies.",
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Configurations/SVT/ADO/ADO.CommonSVTControls.json:48]: user can access the repository and tamper its contents, bypassing any defined policies.  --> user can access the repository, tamper its contents and bypass any defined policies ?

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 4]
```diff
@@ -92,6 +99,146 @@ class CommonSVTControls: ADOSVTBase {
...
+            $inputbody.dataProviderContext.properties.sourcePage.routeValues.Project = $this.ResourceContext.ResourceGroupName;
+            $inputbody.dataProviderContext.properties.permissionSetId = $this.repoPermissionSetId
+            #first check permissions for all branches
+            $inputbody.dataProviderContext.properties.permissionSetToken = "repoV2/$($projectId)/$($this.ResourceContext.ResourceDetails.id)/refs^heads/" 
+            #calling the common method to get permissions object on this level           
+            $excessivePermissionsListOnAllBranches = $this.CheckPermsOnBranch($url,$inputBody,$projectId,$null)
+            if ($null -ne $excessivePermissionsListOnAllBranches -and $excessivePermissionsListOnAllBranches.count -gt 0) {
+                $isControlPassing = $false        
+                $controlResult.AddMessage([VerificationResult]::Failed, "Build service groups have excessive permissions on 'All Branches' level of the repository.");
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:119]: Build service groups  --> Build service accounts

File: src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json [hunk 5]
```diff
@@ -498,6 +498,21 @@
       "Delete repository",
       "Manage permissions"
     ],
+    "BranchesToCheck":[
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json:501]: BranchesToCheck --> BranchesToCheckForExcessivePermissions?

File: src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1 [hunk 6]
```diff
@@ -92,6 +99,146 @@ class CommonSVTControls: ADOSVTBase {
...
+                $formattedGroupsData = $excessivePermissionsListOnAllBranches | Select @{l = 'Group'; e = { $_.Group} }, @{l = 'ExcessivePermissions'; e = { $_.ExcessivePermissions } }
+                $formattedBuildServiceGrpTable = ($formattedGroupsData | Out-String)
+                $controlResult.AddMessage("`nList of groups : `n$formattedBuildServiceGrpTable");
+                $controlResult.AdditionalInfo += "List of excessive permissions on 'All Branches' level:  $($formattedGroupsData).";                       
+            }
+            else {
+                $controlResult.AddMessage("Build service groups do not have excessive permissions on 'All Branches' level of the repository.");
+            }
+            #get all eligible branches from settings and check for each of them
+            $individualBranches = $this.ControlSettings.Repo.BranchesToCheck
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Core/SVT/ADO/ADO.CommonSVTControls.ps1:129]: BranchesToCheckForExcessivePermissions

File: src/AzSK.ADO/Framework/Configurations/SVT/ADO/ADO.CommonSVTControls.json [hunk 7]
```diff
@@ -38,6 +38,23 @@
...
+      "Recommendation": "1. Navigate to Project Settings. --> 2. Click on Repository under Repos. --> 3. Select your repository. --> 4. Click on 'Security'. --> 5. Click on 'All Branches' under 'Git refs permissions'. --> 6. Ensure 'Excessive' permissions of broader groups is not set to 'Allow'. Refer to detailed scan log (Repository.LOG) for broader groups and excessive permissions list. --> 5. Repeat this for any other groups for other individual branches that should not have excessive permissions.",
+      "Tags": [
+        "SDL",
+        "TCP",
+        "Automated",
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Configurations/SVT/ADO/ADO.CommonSVTControls.json:53]: Can  we make it MSW? Because similiar control at repo level is MSW

**Review by arvindsingh-cloud — APPROVED** [review]

**Review by SaikumarRagala — APPROVED** [review]

**Pull request merged**